### ISSUE TITLE
ApiTokenAuthentication 

### ISSUE THREAD

**Issue by Jonatha1983** [mkobit/jenkins-pipeline-shared-libraries-gradle-plugin#82]
Hi @mkobit 

First thing THANKS!!!! You are a genius!!! 

I searched on some docs on how to create:   ApiTokenAuthentication 

(* I am new to Kotlin ) 

Thanks ahead !!!

**Comment by mkobit** [comment]
@Jonatha1983 good call, there is definitely a gap in the documentation around it. I'll try to add some documentation today.

Could you elaborate on what you are trying to accomplish?

**Comment by Jonatha1983** [comment]
We are trying to be able to use the same versions of dependencies used by our Jenknis.  
 I tried something like:
 authentication.set(providers.provider{ BasicAuthentication("user", "pass") })
    baseUrl.set(uri("ourJenknisURL").toURL())

and ran retrieveJenkinsPluginData but got 403 Forbidden with user anonymous

**Comment by Jonatha1983** [comment]
by the way: retrieveJenkinsVersion working perfectly!!

**Comment by mkobit** [comment]
If I remember right, you should be able to replace

```
authentication.set(providers.provider{ BasicAuthentication("user", "pass") })
```

with

```
authentication.set(providers.provider{ ApiTokenAuthentication("user", "apiTokenTextOrReadFromFileOSomewhereElse") })
```

I'm going to add a few tests to verify that that works, but you might not have the appropriate permissions to access them from the Jenkins server. I _think_ you need to have like admin or something type privileges for it, but I'll figure it out some point soon.

**Comment by mkobit** [comment]
I might be wrong on that also - see https://support.cloudbees.com/hc/en-us/articles/218756317-How-can-non-admin-users-view-the-installed-plugins-

I'll do some investigation

**Comment by Jonatha1983** [comment]
I imported your project with include build (Composit Build) to debug it -  and I found: 
1) BasicAuthentication and ApiTokenAuthentication are implemented exactly the same way.
2) Also in both the headers are construct with the word "Authentication" when I changed it to "Authorization" for the ApiTokenAuthentication it works for me.  ( found it here: https://www.baeldung.com/guide-to-okhttp ) 

what do you think ? 

Also I wanted to ask if today the gradle file take in consideration the plugin versions or it just for human validation ? 

Thanks again !!

**Comment by mkobit** [comment]
> BasicAuthentication and ApiTokenAuthentication are implemented exactly the same way.

I _think_ I did that by design, because the same headers were sent - it is tough to remember and I don't have an integration test against a `@JenkinsRule` that sets that up to verify that is the case. I'm hoping to get that done sometime this week.

> Also I wanted to ask if today the gradle file take in consideration the plugin versions or it just for human validation ?

If I am interpreting your question right, the plugins provided to the build script are used in compilation of the `src` and `vars` code (not strict compilation, more of a quick check to verify syntax correctness since Jenkins operates on source code) and are used at runtime in `integrationTest` to be loaded in the `@JenkinsRule` so you can accurately test your shared library against expected Jenkins versions.

**Comment by Jonatha1983** [comment]
Thanks @mkobit !

And about the Authentication vs Authorization ? would you want me to create a pull request with the change ?

**Comment by mkobit** [comment]
@Jonatha1983 ahh, I think I now understand what you are saying. The headers for `ApiTokenAuthentication` should be `Authorization`, not `Authentication`? If so, then yes, would definitely accept an MR for that.

I would to add an additional test that tests those headers against an actual Jenkins instance, but if you don't have time to do that don't worry about it, I can do that part of on my own.

**Comment by mkobit** [comment]
Ok so I went and took a quick look

```
curl localhost:5050/pluginManager/api/json --user admin:11fe35abb2dbfcfc819979ed454bdd06d6
```

Works for me when I was using a local project, but then trying out `ApiTokenAuthentication` wasn't :confused: .

So, probably a bug somewhere in this plugin

I'll look at it more tonight.

**Comment by mkobit** [comment]
> Thanks @mkobit !
> 
> And about the Authentication vs Authorization ? would you want me to create a pull request with the change ?

That is definitely the issue

**Comment by Jonatha1983** [comment]
I am about to push if it is ok the tests i ll push later on this week - since  I need to learn more the code ...

**Comment by mkobit** [comment]
Sounds good @Jonatha1983 , open it whenever you are ready - whatever you have I'll take and get it in, and try and release a new version with the fixes.

I think I'll postpone writing tests against an actual Jenkins in this project since it requires a bit of test infrastructure I don't have yet.

**Comment by Jonatha1983** [comment]
@mkobit  I think you need to gather me permissions

**Comment by mkobit** [comment]
@Jonatha1983 would you mind forking and opening a pull request?

**Comment by Jonatha1983** [comment]
Sure , i ll read how to do it and do it today on.... thanks

**Comment by mkobit** [comment]
Will be released as `0.9.0` as soon as I figure out https://github.com/mkobit/jenkins-pipeline-shared-libraries-gradle-plugin/issues/81

If I don't figure it out by Friday will manually release it instead of relying on Circle CI automation

Closing
